### PR TITLE
fix(core,developer): use `NDEBUG` flag to disable assertions in release build

### PR DESCRIPTION
**Pull request by mcdurdin**
This disables assertions in release builds for all meson-built projects.

See d71cb56ca7bc02a974160f656b9ad32b0846c082 for deeper discussion on the yak shave that impacted all the unit tests.

image

Fixes: #12619
Fixes: #12453 (by implication)

@keymanapp-test-bot skip

### CODE REVIEW

File: common/windows/delphi/general/KeymanPaths.pas [hunk 1]
```diff
@@ -436,12 +436,18 @@ class function TKeymanPaths.RunningFromSource(var keyman_root: string): Boolean;
...
   // within that path then use that as source path
   if TKeymanPaths.RunningFromSource(keyman_root) then
   begin
-    Exit(keyman_root + 'core\build\x86\debug\src\' + Filename);
+{$IFDEF DEBUG}
+    configuration := 'debug';
+{$ELSE}
+    configuration := 'release';
+{$ENDIF}
+    Exit(keyman_root + 'core\build\x86\'+configuration+'\src\' + Filename);
```
Comment (mcdurdin) [inline common/windows/delphi/general/KeymanPaths.pas:450]: Allows us to test both release and debug builds of Core in the Debugger when running within the keymanapp/keyman repo

**Comment by mcdurdin** [comment]
Hmm, unit tests have very unexpectedly failed for [Test: Keyman Core - Windows (Common)](https://build.palaso.org/buildConfiguration/Keyman_Common_KPAPI_TestPullRequests_Windows/506817?buildTab=log&linesState=105&logView=flowAware&focusLine=4369) with this minimal change. More digging needed, tomorrow.

**Comment by ermshiperete** [comment]
Ah, you made the change in `resources/build/meson/standard.meson.build` and so it gets picked up by all the platform builds. I don't think anything else is needed on the Linux side.

**Review by ermshiperete — APPROVED** [review]
LGTM

**Comment by mcdurdin** [comment]
> Ah, you made the change in `resources/build/meson/standard.meson.build` and so it gets picked up by all the platform builds. I don't think anything else is needed on the Linux side.

I constrained the check to msvc though

**Review by rc-swag — DISMISSED** [review]
LGTM

File: core/tests/meson.build [hunk 2]
```diff
@@ -7,7 +7,7 @@
 #
 
 # Note: this version of cmpfiles ignores line endings, which is better for platform independence
-cmpfiles = ['-c', 'import sys; a = open(sys.argv[1], \'r\').read(); b = open(sys.argv[2], \'r\').read(); exit(not (a==b))']
+cmpfiles = ['-c', 'import sys; a = open(sys.argv[1], \'r\').read(); b = open(sys.argv[2], \'r\').read(); sys.exit(not (a==b))']
```
Comment (mcdurdin) [inline core/tests/meson.build:10]: Unrelated ... but maybe due to my python version? In any case `sys.exit` is probably safer

**Comment by mcdurdin** [comment]
The yak shave isn't over:

```
  ../../../src/action.cpp: In function ‘bool km::core::action_item_list_to_actions_object(const km_core_action_item*, km_core_actions*)’:
  ../../../src/action.cpp:64:20: error: variable ‘last_context_item’ set but not used [-Werror=unused-but-set-variable]
     64 |               auto last_context_item = output.back();
        |                    ^~~~~~~~~~~~~~~~~
  ../../../src/action.cpp:74:20: error: variable ‘last_context_item’ set but not used [-Werror=unused-but-set-variable]
     74 |               auto last_context_item = output.back();
        |                    ^~~~~~~~~~~~~~~~~
  cc1plus: all warnings being treated as errors
```

Presumably `last_context_item` is only used in an assertion ... so need to `#ifdef` it away.

**Comment by SabineSIL** [comment]
..changing of assert_failed to test_assert_failed looks good in all files;
... can`t say much about_ ( but it seem OK)
- removing libraries in common/include/test_color.h
- changes in common/windows/delphi/general/KeymanPaths.pas
- using sys.exit instead of exit in core/tests/meson.build

**Comment by darcywong00** [comment]
Is the discussion on https://github.com/keymanapp/keyman/commit/d71cb56ca7bc02a974160f656b9ad32b0846c082 the kind we want to summarize into `/core/docs/internal/ `?

Particularly the fragility of assert() and tests

**Comment by mcdurdin** [comment]
> Is the discussion on [d71cb56](https://github.com/keymanapp/keyman/commit/d71cb56ca7bc02a974160f656b9ad32b0846c082) the kind we want to summarize into `/core/docs/internal/ `?
> 
> Particularly the fragility of assert() and tests

We could add something to our [test wiki](https://github.com/keymanapp/keyman/wiki/Unit-Test-Files-and-Locations); it's not really internal Keyman-relevant documentation, more a trap-for-young-players-in-c++ (i.e. how to use tools properly)

File: core/src/ldml/ldml_markers.cpp [hunk 3]
```diff
@@ -88,7 +92,9 @@ void add_back_markers(std::u32string &str, const std::u32string &src, marker_map
           break;
         } else {
           prepend_marker(str, i->marker, encoding);
+#ifndef NDEBUG
           processed_markers++;
+#endif
```
Comment (darcywong00) [inline core/src/ldml/ldml_markers.cpp:97]: Does l.102 become `test_assert()`?

And does it need to go in a block
```c
#ifndef NDEBUG

#endif
```

Comment (darcywong00) [reply core/src/ldml/ldml_markers.cpp:97]: Similar question for ldml_processor.cpp with lastCtx and its assertions

**Review by darcywong00 — COMMENTED** [review]
questions about ldml_*.cpp

File: core/src/ldml/ldml_markers.cpp (hunk 3 above)
Comment (mcdurdin) [reply core/src/ldml/ldml_markers.cpp:97]: No, because these are not unit test files.

**Review by darcywong00 — APPROVED** [review]
lgtm

**Review by rc-swag — APPROVED** [review]
lgtm 
re-review

**Review by srl295 — APPROVED** [review]

**Pull request merged**